common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -211,63 +214,74 @@ public ManagedBuffer getMergedBlockData(String appId, int shuffleId, int reduceI
-    String indexName = String.format("%s.index", generateFileName(appShuffleId, reduceId));
-    return getFile(appShuffleId.appId, indexName);
+  private File getMergedShuffleIndexFile(String appId, int shuffleId, int reduceId) {
+    String indexName = String.format("%s.index", generateFileName(appId, shuffleId, reduceId));
+    return getFile(appId, indexName);
   }
 
-  private File getMergedShuffleMetaFile(AppShuffleId appShuffleId, int reduceId) {
-    String metaName = String.format("%s.meta", generateFileName(appShuffleId, reduceId));
-    return getFile(appShuffleId.appId, metaName);
+  private File getMergedShuffleMetaFile(String appId, int shuffleId, int reduceId) {
+    String metaName = String.format("%s.meta", generateFileName(appId, shuffleId, reduceId));
+    return getFile(appId, metaName);
   }
 
   @Override
   public String[] getMergedBlockDirs(String appId) {
-    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.get(appId),
+    AppAttemptPathsInfo appAttemptPathsInfo = Preconditions.checkNotNull(appsPathsInfo.get(appId),
       "application " + appId + " is not registered or NM was restarted.");
-    String[] activeLocalDirs = Preconditions.checkNotNull(appPathsInfo.activeLocalDirs,
+    String[] activeLocalDirs = Preconditions.checkNotNull(appAttemptPathsInfo.activeLocalDirs,
       "application " + appId
       + " active local dirs list has not been updated by any executor registration");
     return activeLocalDirs;
   }
 
-  @Override
-  public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
-    logger.info("Application {} removed, cleanupLocalDirs = {}", appId, cleanupLocalDirs);
-    // TODO: [SPARK-33236] Change the message when this service is able to handle NM restart
-    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.remove(appId),
-      "application " + appId + " is not registered or NM was restarted.");
-    Iterator<Map.Entry<AppShuffleId, Map<Integer, AppShufflePartitionInfo>>> iterator =
+  /**
+   * Clean up the AppShufflePartitionInfo for a specific application attempt.
+   * If attemptId is -1, it means to clean up all the AppShufflePartitionInfo from
+   * all the attempts. Otherwise, only the AppShufflePartitionInfo from the specific
+   * application attempt will be cleaned up.
+   */
+  private void cleanupShufflePartitionInfo(String appId, int attemptId) {
+    Iterator<Map.Entry<AppAttemptShuffleId, Map<Integer, AppAttemptShufflePartitionInfo>>> iterator =
       partitions.entrySet().iterator();
     while (iterator.hasNext()) {
-      Map.Entry<AppShuffleId, Map<Integer, AppShufflePartitionInfo>> entry = iterator.next();
-      AppShuffleId appShuffleId = entry.getKey();
-      if (appId.equals(appShuffleId.appId)) {
+      Map.Entry<AppAttemptShuffleId, Map<Integer, AppAttemptShufflePartitionInfo>> entry = iterator.next();
+      AppAttemptShuffleId appAttemptShuffleId = entry.getKey();
+      if (appId.equals(appAttemptShuffleId.appId)
```

Comment:
I think there is a race condition. pushMessage of an older attempt may still be able to create a new appShuffleInfo and partitions may still have an  old appAttemptShuffleId when there is a context switch.

I think we need to also delete such appShuffleInfos during finalize. In finalization, we should remove all the appShuffleInfo that belong to older attempts of the app.